### DISCUSSION TITLE
DPI?

### DISCUSSION THREAD

**Discussion by winny-** [davatorium/rofi#1573] [Q&A]
Hi, having a little trouble understanding Rofi's behavior with DPI.

On my machine I'm able to configure GTK/ & Xft application dpi by setting up my Xresources like so:

```
Xft.dpi: 120
*dpi: 120
```

and Qt apps using:

```
export QT_SCALE_FACTOR=1.22
```

in my shell environment.

Is it possible to use either setting with rofi?  Glanced at the sources, didn't see any logic that feels like the above mechanisms, did I miss something or am misconfiguring rofi?

**Comment by DaveDavenport** [comment]
You can set dpi in config/commandline:
```
      -dpi  number

       Override the default DPI setting.

              • If set to 0, it tries to auto-detect based on X11 screen size (similar to i3 and GTK).

              • If set to 1, it tries to auto-detect based on the size of the monitor that rofi is displayed on (similar to latest Qt 5).

```

If not set it uses the default that pango indicates. 

imho  setting 1 seems to be pretty good in general for my use.   Most themes regrettable uses pixel sizes.
(see output of `rofi -h` to see the detected dpi)

**Reply by winny-** [reply]
Hi thank you for the pointers!

Let me restate what I've learned so far.  Rofi does not use the settings that GTK/Xft or QT does, instead implements it own dpi logic.  I don't really understand how this work- so my statement is probably not quite right.

I did some testing on my 4k 13" laptop display:

1. When I change its resolution to 1080p, the following behavior is observed: (`-dpi 0` -> it appears the dpi of the screen at 4k is still used - rofi is very large, `-dpi 1` -> rofi does detect a sensible lower dpi for a low resolution on a 4k-native monitor.
2. However, it gets more interesting.  When I do the above, then mirror it to an external 1080p tv at 42", I get the following behavior: (`-dpi 0` -> very high dpi like the previous, `-dpi 1` -> extremely low dpi, lower than tested without mirroring)

Version: 1.7.2

I guess this leaves using the dpi setting to set the dpi manually.  I'll take that approach.  Not sure if the above behavior constitutes a bug, but is is a bit confusing.

**Reply by DaveDavenport** [reply]
This is not a lot of info to go on.  I cannot judge if it is a bug based on 'very high'  'extremely low'.
I would need some actual information.

For the different situations can you paste the monitor layout output of `rofi -h` (as mentioned above) 

I use rofi fine with mixed fhd/2k/4k setup and mostly `-dpi 1` detects it nice.

**Reply by DaveDavenport** [reply]
do you use Xorg or Wayland?

**Reply by DaveDavenport** [reply]
I did a quick look at my xps13 with a 4k screen.

Xrandr reports the right monitor height (165 mm), however when reading what X11 reports (used by -dpi 0) it reports 572 mm.

`-dpi 1` gives me a consistent size at different resolutions.

**Comment by WieeRd** [comment]
@DaveDavenport Do you have absolutely no plan to support `Xft.dpi`, `xsettingsd`, or other more generic/conventional means to configure DPI? Although `-dpi 1` can automatically calculate the DPI, the real DPI of a monitor and the actual scaling I want doesn't always align. I use larger scaling than the real DPI on large screens because I'll be sitting further away from them, and smaller scaling on small screens to have more screen real estate.

For this reason, I prefer to keep scaling settings in `~/.Xresources` where I can set machine-specific settings explicitly and exclude it from my dotfiles. ROFI however doesn't follow `Xft.dpi` like the vast majority of GUI apps do, and it's very awkward to maintain a separate dotfiles branch per machine just to adjust ROFI dpi.

**Reply by DaveDavenport** [reply]
Nothing is absolute.

At the moment I have no plans, as I get very little requests for this feature and I already have 0 time to work on things. (see the slow release pace).
A good patch is always welcome.

**Reply by WieeRd** [reply]
Glad to know at least you don't have anything against it.
I'll have to find a time to work on this.

**Reply by JakobKohler** [reply]
I would also enjoy this feature as -dpi 1 scales my font way too big, but -dpi 0 is too small

**Comment by winny-** [comment]
:wave: I no longer use Rofi, but here is how I solved this problem:

1. Set up `config.rasi` to `include` another file, say `~/.cache/dpi.rasi`.
2. In some other script such as .xinitrc or a shell script in your `$PATH` overwrite this other file, say `~/.cache/dpi.rasi`.  You can write out any dpi configuration in this `dpi.rasi` file.
3. You can derive a dpi value from inspecting X Resources DB (`xrdb -query`) or other well known DPI-derived settings like environment `QT_SCALE_FACTOR`.

No longer use rofi or Xorg (much) so I lost these scripts long ago.  Hope this description benefits somebody.

**Reply by WieeRd** [reply]
I'm doing something similar. With `env()` extra file is not required.

```scss
configuration {
    dpi: env(ROFI_DPI, 96);
}
```

Then I set `$ROFI_DPI` in `~/.xsession` script which is not tracked by my dotfiles.

**Reply by DaveDavenport** [reply]
A nice patch would be that you could do:

```css
configuration {
dpi: XRDB(dpi, 120);
}
````
or something similar.
